yarn-project/merkle-tree/src/new_tree.ts

```diff
@@ -15,14 +28,14 @@ import { TreeBase } from './tree_base.js';
  * @returns The newly created tree.
  */
 export async function newTree<T extends TreeBase>(
-  c: new (...args: any[]) => T,
+  c: (db: LevelUp, hasher: Hasher, name: string, depth: number, size: bigint) => T,
   db: LevelUp,
   hasher: Hasher,
   name: string,
   depth: number,
   prefilledSize = 1,
 ): Promise<T> {
-  const tree = new c(db, hasher, name, depth, 0n, undefined);
+  const tree = c(db, hasher, name, depth, 0n);
```

Comment:
So the issue here was:

- newTree and loadTree were called with constructor (typed with new) functions, which all had the same shape before
- now indexed trees have different constructors, because they require the leaf factory and the preimage factory
- Typescript won't let me use the new operator if the function is not typed as a constructor (which makes sense)
- So now newTree and loadTree sometimes need to issue "new" and sometimes not

My solution was ok, let's accept regular functions, and if you are passing a regular constructor just wrap it in a regular function via treeBuilder and you're set. If you are passing a more complex constructor then just wrap it in a closure that provides the factories needed for that specific tree instance.

I tried to type newTree and loadTree with a union type of constructor and regular function, but I didn't find a clean way to narrow the type so it allowed me to do new sometimes and just calling other times